managed/services/uievents/uievents.go

```diff
@@ -0,0 +1,262 @@
+}
+
+// New returns platform Service.
+func New() (*Service, error) {
+	l := logrus.WithField("component", "platform")
+
+	s := Service{
+		l:               l,
+		dashboardUsage:  make(map[string]*DashboardUsageStat),
+		componentsUsage: make(map[string]*ComponentsUsageStat),
+		lastCleanup:     time.Now(),
+	}
+
+	return &s, nil
+}
+
+// ScheduleCleanup schedules internal clean for internal UI event storage.
+func (s *Service) ScheduleCleanup(ctx context.Context) {
+	go func() {
+		ticker := time.NewTicker(cleanupCheckInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				s.cleanupIfNeeded()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (s *Service) FetchMetrics(ctx context.Context, report *telemetry.Config) ([]*pmmv1.ServerMetric_Metric, error) {
+	var metrics []*pmmv1.ServerMetric_Metric
+
+	metrics = s.processDashboardStat(metrics)
+	metrics = s.processComponentsStat(metrics)
+
+	return metrics, nil
+}
+
+func (s *Service) processDashboardStat(metrics []*pmmv1.ServerMetric_Metric) []*pmmv1.ServerMetric_Metric {
+	type Stat struct {
+		TopDashboards         []string `json:"top_dashboards"`
+		SlowDashboardsP95_1s  []string `json:"slow_dashboards_p95_1s"`
+		SlowDashboardsP95_5s  []string `json:"slow_dashboards_p95_5s"`
+		SlowDashboardsP95_10s []string `json:"slow_dashboards_p95_10s"`
+	}
+
+	if len(s.dashboardUsage) > 0 {
+		dashboardStat := &Stat{
+			SlowDashboardsP95_1s:  []string{},
+			SlowDashboardsP95_5s:  []string{},
+			SlowDashboardsP95_10s: []string{},
+		}
+
+		// TopDashboards
+		keys := make([]string, 0, len(s.dashboardUsage))
+		for key := range s.dashboardUsage {
+			keys = append(keys, key)
+		}
+		sort.SliceStable(keys, func(i, j int) bool {
+			return s.dashboardUsage[keys[i]].useCount > s.dashboardUsage[keys[j]].useCount
+		})
+		for i := 0; i < len(keys); i++ {
+			sortedKey := keys[i]
+			stat := s.dashboardUsage[sortedKey]
+			dashboardStat.TopDashboards = append(dashboardStat.TopDashboards, stat.uid)
+		}
+
+		// SlowDashboardsP95
+		sort.SliceStable(keys, func(i, j int) bool {
+			return s.dashboardUsage[keys[i]].loadTime.ValueAtPercentile(95) > s.dashboardUsage[keys[j]].loadTime.ValueAtPercentile(95)
+		})
+		for i := 0; i < len(keys); i++ {
+			sortedKey := keys[i]
+			stat := s.dashboardUsage[sortedKey]
+			p95 := stat.loadTime.ValueAtPercentile(95)
+			if p95 >= 1_000 {
+				dashboardStat.SlowDashboardsP95_1s = append(dashboardStat.SlowDashboardsP95_1s, stat.uid)
+			}
+			if p95 >= 5_000 {
+				dashboardStat.SlowDashboardsP95_5s = append(dashboardStat.SlowDashboardsP95_5s, stat.uid)
+			}
+			if p95 >= 10_000 {
+				dashboardStat.SlowDashboardsP95_10s = append(dashboardStat.SlowDashboardsP95_10s, stat.uid)
+			}
+		}
+
+		marshal, err := json.Marshal(dashboardStat)
+		if err != nil {
+			s.l.Error("failed to marshal to JSON")
+		}
+		metrics = append(metrics, &pmmv1.ServerMetric_Metric{
+			Key:   "ui-events-dashboards",
+			Value: string(marshal),
+		})
```

Comment:
should we append a metric even if the `err` above isn't `nil`

Reply:
good question, I think we need to skip it, there will be nil most likely